rs/execution_environment/src/scheduler/round_schedule.rs

```diff
@@ -229,4 +229,67 @@ impl RoundSchedule {
 
         (canisters_partitioned_by_cores, canisters)
     }
+
+    pub fn finish_canister_execution(
+        canister: &mut CanisterState,
+        fully_executed_canister_ids: &mut Vec<CanisterId>,
+        round_id: ExecutionRound,
+        is_first_iteration: bool,
+        rank: usize,
+    ) {
+        let full_message_execution = match canister.next_execution() {
+            NextExecution::None => true,
+            NextExecution::StartNew => false,
+            // We just finished a full slice of executions.
+            NextExecution::ContinueLong => true,
+            NextExecution::ContinueInstallCode => false,
+        };
+        let scheduled_first = is_first_iteration && rank == 0;
+
+        // The very first canister is considered to have a full execution round for
+        // scheduling purposes even if it did not complete within the round.
+        if full_message_execution || scheduled_first {
+            canister.scheduler_state.last_full_execution_round = round_id;
+
+            // We schedule canisters (as opposed to individual messages),
+            // and we charge for every full execution round.
+            fully_executed_canister_ids.push(canister.canister_id());
+        }
+    }
+
+    pub(crate) fn finish_round(
+        &self,
+        canister_states: &mut BTreeMap<CanisterId, CanisterState>,
+        fully_executed_canister_ids: BTreeSet<CanisterId>,
+    ) {
+        let scheduler_cores = self.scheduler_cores;
+        let number_of_canisters = canister_states.len();
+        let multiplier = (scheduler_cores * number_of_canisters).max(1) as i64;
+
+        // Charge canisters for full executions in the previous round.
+        let mut total_charged_priority = 0;
+        for canister_id in fully_executed_canister_ids {
+            if let Some(canister) = canister_states.get_mut(&canister_id) {
+                total_charged_priority += 100 * multiplier;
+                canister.scheduler_state.priority_credit += (100 * multiplier).into();
+            }
+        }
+
+        // Distribute the charged priority across canisters according to their allocations.
+        let mut total_allocated = 0;
+        for (_canister_id, canister) in canister_states.iter_mut() {
+            let allocated =
+                canister.scheduler_state.compute_allocation.as_percent() as i64 * multiplier;
+            total_allocated += allocated;
+            canister.scheduler_state.accumulated_priority += allocated.into();
+        }
+        // Fully distribute the rest across all canisters.
```

Comment:
Shouldn't the distribution be equal among all canisters? IIUC, we use some uneven distribution based on the ordering of canisters?

Reply:
Discussed offline and fixed in https://github.com/dfinity/ic/pull/1782/commits/5dbdca733d23ad2cecad7b5bf4c81f168e04c929